### PR TITLE
Fix preview of published post in reader mode

### PR DESCRIPTION
**Pull request by ktmn**
Addresses #4656

### CODE REVIEW

File: includes/templates/class-amp-post-template.php [hunk 1]
```diff
@@ -162,7 +162,14 @@ private function set_data() {
 	public function __get( $name ) {
 		switch ( $name ) {
 			case 'post':
-				return get_post( $this->ID );
+				$post = get_post( $this->ID );
+				if ( isset( $_GET['preview_id'] ) && isset( $_GET['preview_nonce'] ) ) {
```
Comment (pierlon) [inline includes/templates/class-amp-post-template.php:166]:
```suggestion
				if ( isset( $_GET['preview_id'], $_GET['preview_nonce'] ) ) {
```

File: includes/templates/class-amp-post-template.php [hunk 2]
```diff
@@ -162,7 +162,14 @@ private function set_data() {
...
+					$id = (int) $_GET['preview_id'];
+					if( wp_verify_nonce( $_GET['preview_nonce'], 'post_preview_' . $id ) ) {
```
Comment (pierlon) [inline includes/templates/class-amp-post-template.php:168]:
```suggestion
					if ( wp_verify_nonce( $_GET['preview_nonce'], 'post_preview_' . $id ) ) {
```

**Review by pierlon — CHANGES_REQUESTED** [review]
Great work here, @ktmn. Thanks for opening this PR!

A few lint errors to be resolved and we should be able to get this merged :smile:.

**Comment by pierlon** [comment]
@ktmn also, don't forget to sign the CLA, as outlined in https://github.com/ampproject/amp-wp/pull/4663#issuecomment-624776452.

**Review by westonruter — CHANGES_REQUESTED** [review]
After digging into this, I found an alternative solution that I think will be more robust. All we need to do is pass in the initial `$post` when constructing the `AMP_Post_Template` and use it from then on. It will have `the_preview` applied already, so we don't need to call it manually:

```diff
--- a/includes/templates/class-amp-post-template.php
+++ b/includes/templates/class-amp-post-template.php
@@ -66,6 +66,13 @@ class AMP_Post_Template {
 	 */
 	public $ID;
 
+	/**
+	 * Post
+	 *
+	 * @var WP_Post
+	 */
+	public $post;
+
 	/**
 	 * AMP_Post_Template constructor.
 	 *
@@ -73,9 +80,11 @@ class AMP_Post_Template {
 	 */
 	public function __construct( $post ) {
 		if ( is_int( $post ) ) {
-			$this->ID = $post;
+			$this->ID   = $post;
+			$this->post = get_post( $post );
 		} elseif ( $post instanceof WP_Post ) {
-			$this->ID = $post->ID;
+			$this->ID   = $post->ID;
+			$this->post = $post;
 		}
 	}
 
@@ -151,22 +160,6 @@ class AMP_Post_Template {
 		$this->data = apply_filters( 'amp_post_template_data', $this->data, $this->post );
 	}
 
-	/**
-	 * Getter.
-	 *
-	 * @since 1.5
-	 *
-	 * @param string $name Property name.
-	 * @return mixed
-	 */
-	public function __get( $name ) {
-		switch ( $name ) {
-			case 'post':
-				return get_post( $this->ID );
-		}
-		return null;
-	}
-
 	/**
 	 * Get template directory for Reader mode.
 	 *
diff --git a/includes/templates/reader-template-loader.php b/includes/templates/reader-template-loader.php
index d518343dc..5833bc9ba 100644
--- a/includes/templates/reader-template-loader.php
+++ b/includes/templates/reader-template-loader.php
@@ -32,5 +32,5 @@ do_action( 'pre_amp_render_post', get_queried_object_id() );
 require_once AMP__DIR__ . '/includes/amp-post-template-functions.php';
 amp_post_template_init_hooks();
 
-$amp_post_template = new AMP_Post_Template( get_queried_object_id() );
+$amp_post_template = new AMP_Post_Template( $post );
 $amp_post_template->load();
```

**Comment by ktmn** [comment]
@westonruter Yeah that makes sense, go with that.

**Comment by westonruter** [comment]
@ktmn cool. Please test the changes in #4665.

**Pull request closed without merging**